README.md
# Isochrone

[![npm version](https://img.shields.io/npm/v/isochrone.svg)](https://www.npmjs.com/package/isochrone)
[![npm downloads](https://img.shields.io/npm/dt/isochrone.svg)](https://www.npmjs.com/package/isochrone)
[![Build Status](https://travis-ci.org/stepankuzmin/node-isochrone.svg?branch=master)](https://travis-ci.org/stepankuzmin/node-isochrone)

Isochrone maps are commonly used to depict areas of equal travel time.
Build isochrones using [OSRM](http://project-osrm.org/), [Turf](http://turfjs.org/) and [concaveman](https://github.com/mapbox/concaveman).

![Screenshot](https://raw.githubusercontent.com/stepankuzmin/galton/master/example.png)

## Installation

```
npm install -g isochrone
```

## Build graph

```shell
wget https://s3.amazonaws.com/metro-extracts.mapzen.com/moscow_russia.osm.pbf
./node_modules/osrm/lib/binding/osrm-extract -p ./node_modules/osrm/profiles/foot.lua moscow_russia.osm.pbf
./node_modules/osrm/lib/binding/osrm-contract moscow_russia.osrm
```

## Usage

```js
const OSRM = require('osrm');
const isochrone = require('isochrone');

const osrm = new OSRM({ algorithm: 'CH', path: './moscow_russia.osrm' });
const startPoint = [37.62, 55.75];

const options = {
  osrm,
  radius: 2,
  cellSize: 0.1,
  intervals: [5, 10, 15]
};

isochrone(startPoint, options)
  .then((geojson) => {
    console.log(JSON.stringify(geojson, null, 2));
  })
  .catch((error) => {
    console.error(error);
  });

```

See [API](https://github.com/stepankuzmin/node-isochrone/blob/master/API.md) for more info.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="3.0.2"></a>
## [3.0.2](https://github.com/stepankuzmin/node-isochrone/compare/v3.0.1...v3.0.2) (2018-09-20)


### Bug Fixes

* fix polygon deintersection ([900e59c](https://github.com/stepankuzmin/node-isochrone/commit/900e59c))



<a name="3.0.1"></a>
## [3.0.1](https://github.com/stepankuzmin/node-isochrone/compare/v3.0.0...v3.0.1) (2018-05-25)


### Bug Fixes

* Add extra catch for empty polygons ([1f4e521](https://github.com/stepankuzmin/node-isochrone/commit/1f4e521))



<a name="3.0.0"></a>
# [3.0.0](https://github.com/stepankuzmin/node-isochrone/compare/v2.0.2...v3.0.0) (2018-04-02)


### Bug Fixes

* **package:** update [@turf](https://github.com/turf)/helpers to version 6.0.1 ([907a4b4](https://github.com/stepankuzmin/node-isochrone/commit/907a4b4))
* **package:** update [@turf](https://github.com/turf)/helpers to version 6.1.0 ([30d51c2](https://github.com/stepankuzmin/node-isochrone/commit/30d51c2))
* **package:** update geojson-rewind to version 0.3.1 ([67efc27](https://github.com/stepankuzmin/node-isochrone/commit/67efc27))


### Features

* update OSRM, drop node v6 support ([d3d05c3](https://github.com/stepankuzmin/node-isochrone/commit/d3d05c3))


### BREAKING CHANGES

* drop node v6 support



<a name="2.0.2"></a>
## [2.0.2](https://github.com/stepankuzmin/node-isochrone/compare/v2.0.1...v2.0.2) (2018-02-05)


### Bug Fixes

* **package:** update [@turf](https://github.com/turf)/bbox to version 4.6.0 ([12efd8c](https://github.com/stepankuzmin/node-isochrone/commit/12efd8c))
* **package:** update [@turf](https://github.com/turf)/bbox to version 4.7.3 ([3365751](https://github.com/stepankuzmin/node-isochrone/commit/3365751))
* **package:** update [@turf](https://github.com/turf)/bbox to version 5.0.4 ([b9171aa](https://github.com/stepankuzmin/node-isochrone/commit/b9171aa))
* **package:** update [@turf](https://github.com/turf)/bbox to version 5.1.0 ([b53faf6](https://github.com/stepankuzmin/node-isochrone/commit/b53faf6))
* **package:** update [@turf](https://github.com/turf)/bbox to version 5.1.5 ([04fed0b](https://github.com/stepankuzmin/node-isochrone/commit/04fed0b))
* **package:** update [@turf](https://github.com/turf)/destination to version 4.6.0 ([caf0fed](https://github.com/stepankuzmin/node-isochrone/commit/caf0fed))
* **package:** update [@turf](https://github.com/turf)/destination to version 4.7.3 ([40c8186](https://github.com/stepankuzmin/node-isochrone/commit/40c8186))
* **package:** update [@turf](https://github.com/turf)/destination to version 5.0.4 ([c0c1ee8](https://github.com/stepankuzmin/node-isochrone/commit/c0c1ee8))
* **package:** update [@turf](https://github.com/turf)/destination to version 5.1.0 ([db65b78](https://github.com/stepankuzmin/node-isochrone/commit/db65b78))
* **package:** update [@turf](https://github.com/turf)/destination to version 5.1.5 ([a866ce1](https://github.com/stepankuzmin/node-isochrone/commit/a866ce1))
* **package:** update [@turf](https://github.com/turf)/helpers to version 4.6.0 ([099e56d](https://github.com/stepankuzmin/node-isochrone/commit/099e56d))
* **package:** update [@turf](https://github.com/turf)/point-grid to version 5.1.0 ([9da0702](https://github.com/stepankuzmin/node-isochrone/commit/9da0702))
* properly handle errors ([43f8532](https://github.com/stepankuzmin/node-isochrone/commit/43f8532))
* **package:** update [@turf](https://github.com/turf)/helpers to version 4.7.3 ([4e27f8c](https://github.com/stepankuzmin/node-isochrone/commit/4e27f8c))
* **package:** update [@turf](https://github.com/turf)/helpers to version 5.0.4 ([38dc2ec](https://github.com/stepankuzmin/node-isochrone/commit/38dc2ec))
* **package:** update [@turf](https://github.com/turf)/helpers to version 5.1.0 ([0f2674f](https://github.com/stepankuzmin/node-isochrone/commit/0f2674f))
* **package:** update [@turf](https://github.com/turf)/helpers to version 5.1.4 ([14def13](https://github.com/stepankuzmin/node-isochrone/commit/14def13))
* **package:** update [@turf](https://github.com/turf)/helpers to version 6.0.0-beta.1 ([2f47fa8](https://github.com/stepankuzmin/node-isochrone/commit/2f47fa8))
* **package:** update [@turf](https://github.com/turf)/helpers to version 6.0.0-beta.4 ([fc1139c](https://github.com/stepankuzmin/node-isochrone/commit/fc1139c))
* **package:** update [@turf](https://github.com/turf)/point-grid to version 4.6.0 ([197ad3f](https://github.com/stepankuzmin/node-isochrone/commit/197ad3f))
* **package:** update [@turf](https://github.com/turf)/point-grid to version 4.7.3 ([9d8c894](https://github.com/stepankuzmin/node-isochrone/commit/9d8c894))
* **package:** update [@turf](https://github.com/turf)/point-grid to version 5.0.0 ([0505a3d](https://github.com/stepankuzmin/node-isochrone/commit/0505a3d))
* **package:** update [@turf](https://github.com/turf)/point-grid to version 5.0.4 ([39407a0](https://github.com/stepankuzmin/node-isochrone/commit/39407a0))
* **package:** update [@turf](https://github.com/turf)/point-grid to version 5.1.5 ([aca7b85](https://github.com/stepankuzmin/node-isochrone/commit/aca7b85))



<a name="2.0.1"></a>
## [2.0.1](https://github.com/stepankuzmin/node-isochrone/compare/v2.0.0...v2.0.1) (2017-07-17)



<a name="2.0.0"></a>
# [2.0.0](https://github.com/stepankuzmin/node-isochrone/compare/v1.0.2...v2.0.0) (2017-07-12)


### Bug Fixes

* **package:** update [@turf](https://github.com/turf)/bbox to version 4.5.1 ([0318cce](https://github.com/stepankuzmin/node-isochrone/commit/0318cce))
* **package:** update [@turf](https://github.com/turf)/destination to version 4.5.1 ([8518f8b](https://github.com/stepankuzmin/node-isochrone/commit/8518f8b))
* **package:** update [@turf](https://github.com/turf)/helpers to version 4.5.1 ([b5fe981](https://github.com/stepankuzmin/node-isochrone/commit/b5fe981))

test/index.js
const path = require('path');
const test = require('tape');
const OSRM = require('osrm');
const area = require('@turf/area').default;
const intersect = require('@turf/intersect').default;
const geojsonhint = require('@mapbox/geojsonhint');
const isochrone = require('../index');

const points = [[7.41337, 43.72956], [7.41375, 43.73339], [7.41862, 43.73216]];

const osrmPath = path.join(__dirname, './data/monaco.osrm');
const osrm = new OSRM({ path: osrmPath });

const options = {
  osrm,
  radius: 2,
  cellSize: 0.1,
  concavity: 2,
  intervals: [5, 10, 15],
  lengthThreshold: 0,
  units: 'kilometers'
};

const timeColors = {
  5: '#0f0',
  10: '#00f',
  15: '#f00'
};

// For argument array intervals: [5, 10, 15, 20]
// Gives: [ [5,10], [5,15], [5,20], [10,15], [10,20], [15,20] ]
const testPairs = options.intervals
  .map(i => ({ head: i, tail: options.intervals.filter(x => x > i) }))
  .map(z => z.tail.map(t => [z.head, t]))
  .reduce((x, y) => x.concat(y), []);

const getIntersectionArea = (a, b) => {
  const intersection = intersect(a, b);
  return intersection ? area(intersection) : 0;
};

test('deintersected isochrone', (t) => {
  t.plan(points.length * (1 + testPairs.length));
  points.forEach(point => isochrone(point, Object.assign({}, options, { deintersect: true }))
    .then((geojson) => {
      geojson.features.forEach((feature) => {
        // eslint-disable-next-line
          feature.properties.fill = timeColors[feature.properties.time];
      });

      const errors = geojsonhint.hint(geojson);
      if (errors.length > 0) {
        errors.forEach(error => t.comment(error.message));
        t.fail('Invalid GeoJSON');
      } else {
        t.pass('Valid GeoJSON');
      }
      // Gives { interval1: isochrone1, interval2: isochrone2, ... }
      const isochrones = options.intervals.reduce(
        (acc, interval) => Object.assign({}, acc, {
          [interval]: geojson.features.find(iso => iso.properties.time === interval)
        }),
        {}
      );
        // test that every smaller isochrone is contained by a larger one
      testPairs.forEach((minutePair) => {
        const [minSmall, minLarge] = minutePair;
        const [small, large] = [isochrones[minSmall], isochrones[minLarge]];
        const areaIntersection = getIntersectionArea(small, large);

        // assert that there is no intersection between isochrones
        if (areaIntersection < 300) {
          t.pass(`Isochrone ${minSmall} does not intersect ${minLarge}`);
        } else {
          t.fail(
            `Isochrone ${minSmall} intersects with ${minLarge} with area ${areaIntersection}`
          );
        }
      });
    })
    .catch(error => t.error(error, 'No error')));
});

test('isochrone', (t) => {
  t.plan(points.length * (1 + testPairs.length));
  points.forEach(point => isochrone(point, Object.assign({}, options, { deintersect: false }))
    .then((geojson) => {
      const errors = geojsonhint.hint(geojson);
      if (errors.length > 0) {
        errors.forEach(error => t.comment(error.message));
        t.fail('Invalid GeoJSON');
      } else {
        t.pass('Valid GeoJSON');
      }
      // Gives { interval1: isochrone1, interval2: isochrone2, ... }
      const isochrones = options.intervals.reduce(
        (acc, interval) => Object.assign({}, acc, {
          [interval]: geojson.features.find(iso => iso.properties.time === interval)
        }),
        {}
      );
        // test that every smaller isochrone is contained by a larger one
      testPairs.forEach((minutePair) => {
        const [minSmall, minLarge] = minutePair;
        const [small, large] = [isochrones[minSmall], isochrones[minLarge]];
        const intersection = intersect(small, large);
        const areaIntersection = intersection ? area(intersection) : 0;
        // assert that the small isochrone overlaps over 90% with the large
        if (areaIntersection / area(small) > 0.9) {
          t.pass(`Isochrone ${minSmall} is contained in ${minLarge}`);
        } else {
          t.fail(`Isochrone ${minSmall} is not contained in ${minLarge}`);
        }
      });
    })
    .catch(error => t.error(error, 'No error')));
});


index.js
const bbox = require('@turf/bbox').default;
const concaveman = require('concaveman');
const deintersect = require('turf-deintersect');
const destination = require('@turf/destination').default;
const helpers = require('@turf/helpers');
const pointGrid = require('@turf/point-grid').default;
const rewind = require('geojson-rewind');

const makeGrid = (startPoint, options) => {
  const point = helpers.point(startPoint);

  const spokes = helpers.featureCollection([
    destination(point, options.radius, 180, options.unit),
    destination(point, options.radius, 0, options.unit),
    destination(point, options.radius, 90, options.unit),
    destination(point, options.radius, -90, options.unit)
  ]);

  const bboxGrid = bbox(spokes);
  const grid = pointGrid(bboxGrid, options.cellSize, { units: options.units });

  return grid.features.map(feature => feature.geometry.coordinates);
};

const groupByInterval = (destinations, intervals, travelTime) => {
  const groups = {};
  intervals.forEach((interval) => {
    const points = destinations
      .filter((point, index) => travelTime[index] !== null && travelTime[index] <= interval * 60)
      .map(d => d.location);

    groups[interval] = points;
  });
  return groups;
};

const makePolygon = (points, interval, options) => {
  const concave = concaveman(points, options.concavity, options.lengthThreshold);
  return helpers.polygon([concave], { time: parseFloat(interval) });
};

const makePolygons = (pointsByInterval, options) =>
  Object.keys(pointsByInterval).reduce((acc, interval) => {
    const points = pointsByInterval[interval];
    if (points.length >= 3) {
      acc.push(makePolygon(points, interval, options));
    }
    return acc;
  }, []);

/**
 * Build isochrone using start point and options
 *
 * @param {Array.<float>} startPoint start point [lng, lat]
 * @param {Object} options object
 * @param {Object} options.osrm - [OSRM](https://github.com/Project-OSRM/osrm-backend) instance
 * @param {number} options.radius - distance to draw the buffer as in
 * [@turf/buffer](https://github.com/Turfjs/turf/tree/master/packages/turf-buffer)
 * @param {number} options.cellSize - the distance across each cell as in
 * [@turf/point-grid](https://github.com/Turfjs/turf/tree/master/packages/turf-point-grid)
 * @param {Array.<number>} options.intervals - intervals for isochrones in minutes
 * @param {boolean} options.deintersect - whether or not to deintersect the isochrones.
 * If this is true, then the isochrones will be mutually exclusive
 * @param {number} [options.concavity=2] - relative measure of concavity as in
 * [concaveman](https://github.com/mapbox/concaveman)
 * @param {number} [options.lengthThreshold=0] - length threshold as in
 * [concaveman](https://github.com/mapbox/concaveman)
 * @param {string} [options.units='kilometers'] - any of the options supported by turf units
 * @returns {Promise} GeoJSON FeatureCollection of Polygons when resolved
 *
 * @example
 * const OSRM = require('osrm');
 * const isochrone = require('isochrone');
 *
 * const osrm = new OSRM({ path: './monaco.osrm' });
 * const startPoint = [7.41337, 43.72956];
 *
 * const options = {
 *   osrm,
 *   radius: 2,
 *   cellSize: 0.1,
 *   intervals: [5, 10, 15],
 *   deintersect: true
 * };
 *
 * isochrone(startPoint, options)
 *   .then((geojson) => {
 *     console.log(JSON.stringify(geojson));
 *   });
 */
const isochrone = (startPoint, options) =>
  new Promise((resolve, reject) => {
    try {
      const endPoints = makeGrid(startPoint, options);
      const coordinates = [startPoint].concat(endPoints);
      options.osrm.table({ sources: [0], coordinates }, (error, table) => {
        if (error) {
          reject(error);
        }

        const travelTime = table.durations[0] || [];

        try {
          const pointsByInterval = groupByInterval(table.destinations, options.intervals, travelTime);
          const polygons = makePolygons(pointsByInterval, options);

          const features = options.deintersect ? deintersect(polygons) : polygons;
          const featureCollection = rewind(helpers.featureCollection(features));

          resolve(featureCollection);
        }
        catch (e) {
          reject(e);
        }
      });
    } catch (e) {
      reject(e);
    }
  });

module.exports = isochrone;
